proposals/record-structs.md

```diff
@@ -0,0 +1,301 @@
+    }
+    public static bool operator==(R1 r1, R1 r2)
+        => r1.Equals(r2);
+    public static bool operator!=(R1 r1, R1 r2)
+        => !(r1 == r2);    
+    public override int GetHashCode()
+    {
+        return Combine(
+            EqualityComparer<T1>.Default.GetHashCode(P1),
+            EqualityComparer<T2>.Default.GetHashCode(P2));
+    }
+}
+```
+
+### Printing members: PrintMembers and ToString methods
+
+The record struct includes a synthesized method equivalent to a method declared as follows:
+```C#
+private bool PrintMembers(System.Text.StringBuilder builder);
+```
+
+The method does the following:
+1. for each of the record struct's printable members (non-static public field and readable property members), appends that member's name followed by " = " followed by the member's value: `this.member`, separated with ", ",
+2. return true if the record struct has printable members.
+
+The `PrintMembers` method can be declared explicitly.
+It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+The record struct includes a synthesized method equivalent to a method declared as follows:
+```C#
+public override string ToString();
+```
+
+The method can be declared explicitly. It is an error if the explicit declaration does not match the expected signature or accessibility.
+
+The synthesized method:
+1. creates a `StringBuilder` instance,
+2. appends the record struct name to the builder, followed by " { ",
+3. invokes the record struct's `PrintMembers` method giving it the builder, followed by " " if it returned true,
+4. appends "}",
+5. returns the builder's contents with `builder.ToString()`.
+
+For example, consider the following record struct:
+
+``` csharp
+record struct R1(T1 P1, T2 P2);
+```
+
+For this record struct, the synthesized printing members would be something like:
+
+```C#
+struct R1 : IEquatable<R1>
+{
+    public T1 P1 { get; init; }
+    public T2 P2 { get; init; }
+
+    protected virtual bool PrintMembers(StringBuilder builder)
+    {
+        builder.Append(nameof(P1));
+        builder.Append(" = ");
+        builder.Append(this.P1); // or builder.Append(this.P1); if P1 has a value type
+        builder.Append(", ");
+
+        builder.Append(nameof(P2));
+        builder.Append(" = ");
+        builder.Append(this.P2); // or builder.Append(this.P2); if P2 has a value type
+
+        return true;
+    }
+
+    public override string ToString()
+    {
+        var builder = new StringBuilder();
+        builder.Append(nameof(R1));
+        builder.Append(" { ");
+
+        if (PrintMembers(builder))
+            builder.Append(" ");
+
+        builder.Append("}");
+        return builder.ToString();
+    }
+}
+```
+
+## Positional record struct members
+
+In addition to the above members, record structs with a parameter list ("positional records") synthesize
+additional members with the same conditions as the members above.
```

Comment:
I don't know what this means. What conditions? Some of the conditions apply, some do not, and you don't reference a specific section so I don't really know what this means.

Reply:
This is from the original records spec. I think it refers to the intro blob:
```
Members are synthesized unless a member with a "matching" signature is declared in the record struct body or
an accessible concrete non-virtual member with a "matching" signature is inherited.
```

Note: I've kept the inheritance here, since technically ValueType can be your own implementation.

---
In reply to: [553002512](https://github.com/dotnet/csharplang/pull/4307#discussion_r553002512) [](ancestors = 553002512)